common/src/api/internal/nexus.rs

```diff
@@ -103,3 +103,25 @@ pub struct UpdateArtifact {
 pub enum UpdateArtifactKind {
     Zone,
 }
+
+/// Sent to a sled agent to request the contents of an Instance's serial console.
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq)]
+pub struct InstanceSerialConsoleRequest {
+    /// Character index in the serial buffer from which to read.
+    /// - If positive, this is the number of bytes output since instance start.
+    /// - If negative, this indexes backwards from the end of the most recently buffered data.
+    pub byte_offset: Option<isize>,
```

Comment:
I'm trying to wrap my head around the different uses of this request structure, and I think I'm struggling :sweat_smile: 

- What would `None` mean here?
- In the positive vs negative cases, we are using two entirely different frames-of-reference, right? Like, "0" always means "the first byte emitted from the instance", but "-1" means "the most recent byte we've seen" (which *could* be equivalent to a positive value `byte_offset`, depending on what has been buffered)? I'm a little surprised by this non-linear indexing usage.

Does an `Option<isize>` actually make sense for this dual-reference indexing? WDYT about something like this:

```rust
enum Offset {
  // Index from the start of the instance.
  FromStart(usize),

  // Return the most recent bytes seen from the instance's serial console.
  MostRecent(usize),
}
```

Reply:
I'm definitely in favor of doing this in the underlying code, I was kind of inclined to avoid it at the API surface itself for https://github.com/oxidecomputer/omicron/issues/117 reasons

Reply:
Enums like this aren't really a problem. The problem mostly arises in enums that mix string types with richer data structures. This is less of an issue today now that we have our own generator than it was initially though, so I think it's fine either way. cc @david-crespo to check me on that. 

In general designing the api to be more expressive and explicit is probably better from a usability perspective.

Reply:
Yep, that's correct. I think the enum would turn out fine. I believe in the TS client it would turn into something very usable like this:

```ts
type Offset = 
 | { type: "FromStart"; value: number }
 | { type: "MostRecent"; value: number }
```

This is partly because our generator is better than the the off the shelf options for our purposes, but I also think Dropshot's OpenAPI output has been improved.

Reply:
So I've been running into https://github.com/oxidecomputer/dropshot/blob/784399af2473906f96e2c49698c3fd94933c57b7/dropshot/src/type_util.rs#L71 when I put the tagged enum straight into the endpoint - am I missing something obvious?

Reply:
Not sure without seeing how you're doing it, but here are a couple that work:

https://github.com/oxidecomputer/omicron/blob/e2bd1e85ccad60ff9b58c751929fa2c51ce9b6b7/nexus/src/external_api/params.rs#L331-L358

https://github.com/oxidecomputer/omicron/blob/e2bd1e85ccad60ff9b58c751929fa2c51ce9b6b7/sled-agent/src/params.rs#L28-L36

Reply:
@lifning - if you have a branch that includes the enum (even if it makes dropshot unhappy), I'd be happy to help debug?

Reply:
https://github.com/lifning/omicron/tree/wip is where my branch is, if you wanna take a look

Reply:
Hey, so I checked that branch out - I'm seeing errors while building, because there's a dependency from `omicron-sled-agent` --> `propolis-client`, which appears to be pointing to a revision of propolis that does not exist.

See: https://github.com/lifning/omicron/blame/wip/sled-agent/Cargo.toml#L27

This change looks like it comes from https://github.com/lifning/omicron/commit/0936a92aa4fb9b3cc133415e2f1ca836ac8b1825#diff-46cfcab5a82aa792a22470d4b5f050a5d7c9618ec8c44971ab7cde19911fe4ff 

Would it be a problem to just use the latest revision of Propolis here?

Reply:
Ah, I think this may be the dependency on https://github.com/oxidecomputer/propolis/pull/131

Reply:
Yeah--is it okay to merge that ahead of this being ready to go in to actually use it?

Reply:
Yea, I'd say go for it. It's small and self-contained. You can still make changes after that as needed.

Reply:
Now that I understand [the issue](https://github.com/oxidecomputer/dropshot/issues/369) better, I think the problem is that this about a query string, and Dropshot doesn't know how you would represent this enum as a query string. When I agreed the enum was fine, I thought we were talking about a POST request body or a response body. Sorry!

In order to make this work in the query string it would need to look more like a simple struct (like it was before haha):

```rs
struct QueryArgs {
  from_start_offset: Option<usize>,
  from_end_offset: Option<usize>
}
```

so it can get turned into `?from_start_offset=100` etc.

Reply:
luckily enough, that's the first thing i did before spending a couple days banging my head on getting it to accept the enum it -- so i simply dropped that commit from my branch :)